mkosi/__init__.py

```diff
@@ -3543,14 +3537,15 @@ def run_verb(args: Args, images: Sequence[Config]) -> None:
 
     # First, process all directory removals because otherwise if different images share directories a later
     # image build could end up deleting the output generated by an earlier image build.
+    tools: Optional[Config]
 
     for config in images:
         if not args.verb.needs_build() and args.verb != Verb.clean:
             continue
 
-        if config.tools_tree and config.tools_tree.name == "default":
-            with finalize_default_tools(args, config) as tools:
-                fork_and_wait(lambda: run_clean(args, tools)) # pyright: ignore
+        if config.tools_tree and config.tools_tree == Path("default"):
+            tools = finalize_default_tools(args, config, resources=resources)
+            fork_and_wait(lambda: run_clean(args, tools)) # type: ignore
```

Comment:
```suggestion
            cleanup_tools = finalize_default_tools(args, config, resources=resources)
            fork_and_wait(lambda: run_clean(args, cleanup_tools))
```
The reason mypy complains about this now is because they try to reify the type for the whole function, which makes the possible `None` further down travel backward in time to this point. Changing the name allows this instance of `tools` to have a different type. This also make either type ignore comment unnecessary.